File: code/004/go/main.go
/* From Python to Go: Python: 004 - lists */

package main

// Import
import (
	"fmt"
	"os"
	"slices"
	"strings"
)

// Aux functions
func getAutomationInput() string {
	/* Helper function to get the automation input */
	return os.Getenv("AUTOMATION_INPUT")
}

// Main functions
func main() {
	// Get the input
	automationInput := getAutomationInput()
	fmt.Println(automationInput)

	// Create a slice
	automationSlice := strings.Split(automationInput, ",")
	fmt.Println(automationSlice)

	// Add new element to the end of the slice
	automationSlice = append(automationSlice, "new_device")
	fmt.Println(automationSlice)

	// Add new element to the beginning of the slice
	automationSlice = slices.Insert(automationSlice, 0, "provisioning_required")
	fmt.Println(automationSlice)

	// Check if the element exist in the slice
	if slices.Contains(automationSlice, "provisioning_required") {
		fmt.Printf("provisioning_required is for the device %v\n", automationSlice[0])
	}

	// Remove elemement from the slice and by index
	deleteIndex := slices.Index(automationSlice, "new_device")
	automationSlice = append(automationSlice[:deleteIndex], automationSlice[deleteIndex+1:]...)
	fmt.Println(automationSlice)

	// Change element
	automationSlice[0] = "provisioning_done"
	fmt.Println(automationSlice)

	// Sort the slice
	slices.Sort(automationSlice)
	fmt.Println(automationSlice)

	// Reverse the slice
	slices.Reverse(automationSlice)
	fmt.Println(automationSlice)

	// Merge list into the slice
	automationString := strings.Join(automationSlice, ",")
	fmt.Println(automationString)
}

File: code/004/python/main.py
"""From Python to Go: Python: 004 - lists and slices"""

# Import part
import os


# Functions
def get_automation_input() -> str:
    """Helper function to get the automation input"""
    return os.getenv("AUTOMATION_INPUT")


# Read environmental value
automation_input = get_automation_input()
print(f"{automation_input=}")

# Create a list from string
automation_list = automation_input.split(",")
print(f"{automation_list=}")

# Add an element to the end of the list
automation_list.append("new_device")
print(f"{automation_list=}")

# Add an element to the beginning of the list
automation_list.insert(0, "provisioning_required")

# Check if that element is in the list
if "provisioning_required" in automation_list:
    print(f"provisioning_required is for the device {automation_list[1]}")

# Remove an element from the list
automation_list.remove("new_device")
print(f"{automation_list=}")

# Change the element in the list
automation_list[0] = "provisioning_done"
print(f"{automation_list=}")

# Remove the element by index
del automation_list[0]
print(f"{automation_list=}")

# Sort the list
automation_list.sort()
print(f"{automation_list=}")

# Reverse the list
automation_list.reverse()
print(f"{automation_list=}")

# Merger list into a string
automation_string = ",".join(automation_list)
print(f"{automation_string=}")
